### PR TITLE
Augment emit_nvtx to help connect backward-pass Function apply calls with their corresponding forward pass ops

### PR DESCRIPTION
**Pull request by mcarilli**
Often, we find ourselves looking at some long-running kernel or emit_nvtx range on an nvvp profile and trying to connect it to the offending line in a training script.  If the op is in the forward pass that's easy:  ops are enqueued explicitly from the Python side, so tracking it down with manual nvtx ranges supplemented by the built-in emit_nvtx ranges is straightforward.  If the op is in the backward pass, it's much more difficult.  From the Python side, all you can do is wrap loss.backward() in an nvtx range, and if you also use emit_nvtx, the automatic ranges provide only local information.  Right now, the only consistent way to connect backward-pass kernels to their associated forward-pass lines of Python is to understand your script line by line, and know exactly where in the backward pass you are.

This PR augments the existing nvtx machinery to bridge the gap between forward and backward, allowing connection of backward-pass Function apply calls to the forward-pass operations that required/created those Functions.

The method is simple and surgical.  During the forward pass, when running with emit_nvtx, the nvtx range for each function in VariableType is tagged with the current sequence number.  During the backward pass, the nvtx range associated with each Function's operator() is tagged with that Function's stashed sequence number, which can be compared to "current sequence numbers" from the forward pass to locate the associated op.

Double-backward is not a problem.  If a backward pass with create_graph = True is underway, the relationship between backward and double-backward is conceptually the same as the relationship between forward and backward:  The functions in VariableType still spit out current-sequence-number-tagged ranges, the Function objects they create still stash those sequence numbers, and in the eventual double-backward execution, their operator() ranges are still tagged with the stashed numbers, which can be compared to "current sequence numbers" from the backward pass.

Minor caveats:  

- The sequence number is thread-local, and many VariableType functions (specifically, those without a derivative explicitly defined in derivatives.yaml) don't create an associated function object (instead delegating that to sub-functions further down the call chain, perhaps called from within at::native functions that route back through VariableType by calling at::function_name).  So the correspondence of stashed sequence numbers in Function operator() ranges with numbers in forward-pass ranges is not guaranteed to be 1 to 1.  However, it's still a vast improvement over the current situation, and I don't think this issue should be a blocker.
- Feel free to litigate my use of stringstream in profiler.cpp.  I did it because it was easy and clean.  If that's too big a hammer, let's figure out something more lightweight.

### CODE REVIEW

**Comment by ezyang** [comment]
Tests need to be updated for the change.

File: torch/csrc/autograd/profiler.cpp [hunk 1]
```diff
@@ -95,14 +97,25 @@ RecordFunction::RecordFunction(const char* name) {
   pushRange(name);
 }
 
+RecordFunction::RecordFunction(const char* name, int64_t current_sequence_nr) 
+{
+  if (state == ProfilerState::Disabled)
+    return;
+  std::stringstream s;
+  s << name << ", current sequence nr " << current_sequence_nr;
+  pushRange(s.str());
```
Comment (apaszke) [inline torch/csrc/autograd/profiler.cpp]: All string formatting should happen when the trace is processed. Right now it happens in the hot path and greatly increases the overhead of the profiler.

**Review by apaszke — CHANGES_REQUESTED** [review]

File: torch/csrc/autograd/profiler.cpp (hunk 1 above)
Comment (mcarilli) [reply torch/csrc/autograd/profiler.cpp]: Again, thank you for the quick review.

Less verbose is definitely an improvement.  How about `Add, seq=123` and `AddBackward, forward_seq=123`?

`All string formatting should happen when the trace is processed.` what do you mean by this?  `trace` is an overloaded/ambiguous term here.  Should I defer the string processing to [the actual nvtxRangePushA call`](https://github.com/pytorch/pytorch/blob/ca57b19fd7ee94f62e9c9aae97a7687f257caa29/torch/csrc/autograd/profiler.cpp#L48), which is the latest it can be deferred?

The id needs to be dynamically appended to a string somehow, some way.  Globally, I don't see how this overhead can be avoided.

**Comment by mcarilli** [comment]
@apaszke I've updated the PR to move all string processing so that it only occurs in the narrowest NVTX-specific case, with less verbose annotations (`seq=N` and `stashed seq=N` for VariableType::function and Function object operator() ranges respectively).  Again, I'm not sure if this is exactly what you wanted, but if not, please clarify so we can iterate on it.

To restrict string processing to the NVTX-specific case, I had to expand the function signature of `profiler::pushRange` with some defaulted arguments, but those additional arguments may turn out to be useful for other cases in the future, so I don't think they do any harm. 

@ezyang Now that the annotation changes have been restricted to NVTX only, the existing `test_profiler` in `test_autograd.py` passes, because it doesn't cover NVTX.  I'm not sure how to add an NVTX-specific test, short of running the entire test script through nvprof and loading and inspecting the resulting nvprof file.

I've also updated the docstring of `emit_nvtx` in `torch/autograd/profiler.py` with a comprehensive and hopefully clear explanation of how the sequence numbers

**Comment by apaszke** [comment]
@mcarilli I don't see why can't we postpose the string processing until after the tracer finishes. You can replace the event list to hold a struct containing both the name of the range, and (optionally) its sequence number, and stitch them together only right before you return them to Python.

**Comment by mcarilli** [comment]
@apaszke Again, I'm not sure what you mean.  My goal is to create NVTX ranges annotated with sequence numbers, and NVTX ranges are pushed and popped on the fly (wrapping the calls they are intended to mark).  They don't return anything to Python, they end up in the sqlite file created by the nvprof executable.  I don't believe there is a way to stitch sequence number information onto NVTX ranges after they've been pushed and popped.  

If you're suggesting I move the annotations to use profiler::Event instead of NVTX, I agree that string processing could be postponed, but I'm afraid that defeats the purpose for us.  The research and arch people I'm supporting need sequence number information to be integrated into the nvprof dumps.

**Comment by apaszke** [comment]
@mcarilli Ok, now that the code is restricted to the NVTX path only (originally it hasn't been), I think we can accept this patch. However, I still think there are a few things that might make the profiler less useful for us, in particular the unnecessary use of `std::string msg`, which could easily have been a `const char *` that would save us on allocations and destructor calls in all other paths. I tried using the profiler yesterday for checking JIT perf and I have concluded that it's no longer as useful as it used to, due to all overheads that have creeped up over the past year. I'm trying to see if we can push them down again, but any such patch unfortunately makes them worse.

I'm surprised people are using the NVTX mode this much. I haven't found it very useful due to how slow NVTX is.

File: torch/csrc/autograd/profiler.cpp [hunk 2]
```diff
@@ -37,13 +39,19 @@ void mark(std::string name, bool include_cuda /* = true */) {
   }
 }
 
-void pushRange(std::string name) {
+void pushRange(std::string name, std::string msg/*= {}*/, int64_t sequence_nr/*= -1*/) {
```
Comment (apaszke) [inline torch/csrc/autograd/profiler.cpp]: Can you please make `msg` a `const char *`?

**Review by apaszke — COMMENTED** [review]
LGTM, but please let's avoid creating strings too much

File: torch/csrc/autograd/profiler.cpp (hunk 2 above)
Comment (mcarilli) [reply torch/csrc/autograd/profiler.cpp]: Yeah, our people need their NVTX...

I'll push this later today.  I will also think about further optimizations (for example, there's no reason nested ranges in the forward pass need to append the sequence number except at the top level).

Thanks!

**Comment by mcarilli** [comment]
@apaszke I've pushed a new commit with the requested changes, that also cleans up my profiler.py docstrings so they should pass the flake8.  Thank you for your help!

**Review by apaszke — APPROVED** [review]
